api/cache/objectslist.go

```diff
@@ -0,0 +1,113 @@
+package cache
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/bluele/gcache"
+	cid "github.com/nspcc-dev/neofs-api-go/pkg/container/id"
+	"github.com/nspcc-dev/neofs-api-go/pkg/object"
+)
+
+/*
+	This is an implementation of a cache which keeps unsorted lists of objects' IDs (all versions)
+	for a specified bucket and a prefix.
+
+	The cache contains gcache whose entries have a key: ObjectsListKey struct and a value: list of ids.
+	After putting a record it lives for a while (default value is 60 seconds).
+
+	When we receive a request from the user we try to find the suitable and non-expired cache entry, go through the list
+	and get ObjectInfos from common object cache or with a request to NeoFS.
+
+	When we put an object into a container we invalidate entries with prefixes that are prefixes of the object's name.
+*/
+
+type (
+	// ObjectsListCache provides interface for cache of ListObjectsV2 in a layer struct.
+	ObjectsListCache interface {
+		Get(key ObjectsListKey) []*object.ID
+		Put(key ObjectsListKey, oids []*object.ID) error
+		CleanCacheEntriesChangedWithPutObject(objectName string, cid *cid.ID)
```

Comment:
The logic for DELETE will be the same. Should we give this function a shorter and more generic name?

Reply:
No, I decided not to invalidate lists in DELETE because every time, when we go through a list of object ids, we make a request to the object's cache (as expected, we delete the object's header from cache after DELETE request from user) and after we got nil, go to NeoFS and get "object not found".  And then we delete this oid from the list. 
```
obj := n.objectFromObjectsCacheOrNeoFS(ctx, bkt.CID, ids[i])
		if obj == nil {
			// mark ids[i] as an address of invalid object
			ids[i] = nil
			continue
		}
```